src/main/java/uk/gov/pay/api/exception/mapper/ViolationExceptionMapper.java

```diff
@@ -0,0 +1,71 @@
+package uk.gov.pay.api.exception.mapper;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.dropwizard.jersey.validation.JerseyViolationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.pay.api.model.PaymentError;
+
+import javax.annotation.Priority;
+import javax.validation.ConstraintViolation;
+import javax.validation.ConstraintViolationException;
+import javax.validation.Path;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.Optional;
+
+import static uk.gov.pay.api.model.PaymentError.Code.CREATE_PAYMENT_VALIDATION_ERROR;
+
+@Priority(1)
+public class ViolationExceptionMapper implements ExceptionMapper<JerseyViolationException> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ViolationExceptionMapper.class);
+
+    @Override
+    public Response toResponse(JerseyViolationException exception) {
+        LOGGER.error(exception.getMessage());
+        ConstraintViolation<?> firstException = getFirstException(exception);
+        String message = firstException.getMessage();
+        String fieldName = getApiFieldName(firstException);
+        PaymentError paymentError = PaymentError.aPaymentError(fieldName, CREATE_PAYMENT_VALIDATION_ERROR, message);
+
+        return Response.status(422)
+                .entity(paymentError)
+                .build();
+    }
+
+    private String getApiFieldName(ConstraintViolation<?> firstException) {
+        Field field = getField(firstException);
+        return getJacksonPropertyName(field).orElse(field.getName());
+    }
+
+    private Optional<String> getJacksonPropertyName(Field field) {
+        return Arrays.stream(field.getAnnotations())
+                .filter(annotation -> annotation instanceof JsonProperty)
+                .findFirst()
+                .map(annotation -> ((JsonProperty) annotation).value());
+    }
+    
+    private Field getField(ConstraintViolation<?> firstException) {
+        Class<?> leafBean = firstException.getLeafBean().getClass();
+        String fieldName = getLastElementOfPath(firstException.getPropertyPath());
+        try {
+            return leafBean.getField(fieldName);
+        } catch (NoSuchFieldException e) {
+            LOGGER.error(String.format("Field %s does not exist on class %s", fieldName, leafBean.toString()));
+            throw new WebApplicationException(e);
+        }
+    }
+    
+    private String getLastElementOfPath(Path path) {
+        String[] pathParts = path.toString().split("\\.");
+        return pathParts[pathParts.length - 1];
+    }
+
+    private ConstraintViolation<?> getFirstException(ConstraintViolationException exception) {
+        return exception.getConstraintViolations().iterator().next();
```

Comment:
do you really need a method for this line?

Reply:
Trying out ideas from reading 'Clean Code' which promotes the idea of small methods with everything in a method being of a similar logical level. Since the contents of `getFirstException` deals with `iterator().next()` its a level much lower than the other code within `toResponse()` and I believe it keeps reading `toResponse` cleaner, so I think it makes sense to put in it's own method. Unless you feel strongly I'd like to keep it but welcome the discussion.

Reply:
I do agree with that idea, especially when the method names describe what is being done, e.g. `getField` and `getJacksonPropertyName` but my view is that if something is perfectly readable, extracting into methods can lead to unreadability. `ConstraintViolation<?> firstException = exception.getConstraintViolations().iterator().next();` isn't hard to understand at all. It's like saying `String s = list.get(0)` can be extracted to `String s = getFirstElement(list)`...

Reply:
Ok I've removed the method. I still like the idea of keeping the contents of methods at the same conceptual level and even though something maybe simple to understand, as it is in this case, it can still add to the clutter and cognitive load when first understanding what a method is doing. Using even small single line methods keeps the lower level implementation available for any reader who wishes to know more. Maybe its a case of when we're reviewing we need to know exactly how something works and so want to see the implementation immediately and not dig deeper for it, whilst a later reader might be happy to know the flow of a given method and go deeper as and when needed.

I entirely agree with your point about needlessly heavily nested methods diminishing readability and I've felt this frustration myself, so on balance and since you feel strongly I have removed it, thanks for the discussion, it certainly helps me be more considerate of how people read my code and hopefully strike a better balance 👍